### PR TITLE
pyspark metrics:

### PR DESCRIPTION
**Pull request by dvoykov**
- implementations for CountFewLastDayRows, CheckAdversarialValidation
 - rework for CountNull, CountLastDayRows
 - changed test results for CheckAdversarialValidation

### CODE REVIEW

File: pure/metrics.py [hunk 1]
```diff
@@ -1584,30 +1613,51 @@ def __post_init__(self):
...
+
+        is_similar = True
+        importance_dict = {}
+
+        # cross validation, binary classifier
+        classifier = RandomForestClassifier(random_state=42)
+        cv_result = cross_validate(classifier, X, y, cv=5, scoring='roc_auc', return_estimator=True)
+        mean_score = np.mean(cv_result['test_score'])
+
+        # columns differences/importances
```
Comment (IgnatovN) [inline pure/metrics.py:1639]: Убрать комментарий

File: pure/metrics.py [hunk 2]
```diff
@@ -1584,30 +1613,51 @@ def __post_init__(self):
...
+            importances = np.mean(
+                [est.feature_importances_ for est in cv_result['estimator']],
+                axis=0
+            )
+            importance_dict = dict(zip(columns, np.around(importances, 5)))
+
+        return is_similar, importance_dict, np.around(mean_score, 5)
+
+    def _call_pandas(self, df: pd.DataFrame) -> Dict[str, Any]:
         # check for existence of numeric values
```
Comment (IgnatovN) [inline pure/metrics.py:1652]: Убрать комментарий

**Review by IgnatovN — CHANGES_REQUESTED** [review]

File: pure/metrics.py [hunk 3]
```diff
@@ -1521,8 +1530,28 @@ def _call_pandas(self, df: pd.DataFrame) -> Dict[str, Any]:
...
-        # TODO: add pyspark implementation of call method
-        raise NotImplementedError("This method is not implemented yet.")
+        from pyspark.sql.functions import col, asc, to_date, isnan
+
+        empty_rows_qty = df.filter(col(self.column).isNull() | isnan(self.column)).count()
+
+        if empty_rows_qty != 0:
+            raise ValueError(f"None/nan values in column: {self.column}.")
+
+        # asc sorted daily rows qty
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py [hunk 4]
```diff
@@ -1584,30 +1613,51 @@ def __post_init__(self):
...
+            (self._start_1 < self._end_2 < self._end_1):
             raise ValueError("Slices must not overlap.")
 
+    def _compare_samples(self, X: np.ndarray, y: np.ndarray, columns: List[str]) ->\
+        Tuple[bool, Dict[str, float], float]:
+
+        is_similar = True
+        importance_dict = {}
+
+        # cross validation, binary classifier
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py (hunk 1 above)
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py [hunk 5]
```diff
@@ -1622,33 +1672,70 @@ def _call_pandas(self, df: pd.DataFrame) -> Dict[str, Any]:
...
+            raise ValueError("Dataframe contains only non-numeric values.")
+
+        num_data = df.select(num_cols)
+
+        # fill na values by global minimum minus 1000
+        col_mins = num_data.agg(*[min_(col(column)).alias(f"{column}") for column in num_cols])
+        min_value = col_mins.select(least(*col_mins.columns)).collect()[0][0]
+        num_data = num_data.fillna(min_value - 1000)
+
+        # try to slice df
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py [hunk 6]
```diff
@@ -1584,30 +1613,51 @@ def __post_init__(self):
...
-        start_2, end_2 = self.second_slice[0], self.second_slice[1]
+        self._start_1, self._end_1 = self.first_slice[0], self.first_slice[1]
+        self._start_2, self._end_2 = self.second_slice[0], self.second_slice[1]
 
         # check for start >= end
-        if start_1 >= end_1 or start_2 >= end_2:
+        if self._start_1 >= self._end_1 or self._start_2 >= self._end_2:
             raise ValueError("First value in slice must be lower than second value in slice.")
 
         # check for slices overlap
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py (hunk 2 above)
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py [hunk 7]
```diff
@@ -1622,33 +1672,70 @@ def _call_pandas(self, df: pd.DataFrame) -> Dict[str, Any]:
...
+
+        # check for existence of numeric values
+        num_types = ['int', 'bigint', 'float', 'double']
+        num_cols = [col for col, data_type in df.dtypes if data_type in num_types]
+        if len(num_cols) == 0:
+            raise ValueError("Dataframe contains only non-numeric values.")
+
+        num_data = df.select(num_cols)
+
+        # fill na values by global minimum minus 1000
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

**Review by IgnatovN — APPROVED** [review]

File: pure/metrics.py [hunk 8]
```diff
@@ -181,20 +181,20 @@ def _call_pandas(self, df: pd.DataFrame) -> Dict[str, Any]:
...
-            if self.aggregation == "any":
-                mask |= col(column).isNull() | isnan(col(column))
-            elif self.aggregation == "all":
-                mask &= col(column).isNull() | isnan(col(column))
-            else:
-                raise ValueError("Unknown value for aggregation")
-
-        k = df.select(count(when(mask, column))).collect()[0][0]
+
+        # empty(nan, null) columns count for each row
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py [hunk 9]
```diff
@@ -1584,30 +1613,51 @@ def __post_init__(self):
...
+
+        return is_similar, importance_dict, np.around(mean_score, 5)
+
+    def _call_pandas(self, df: pd.DataFrame) -> Dict[str, Any]:
         # check for existence of numeric values
         num_data = df.select_dtypes(include=["number"])
         if len(num_data) == 0:
             raise ValueError("Dataframe contains only non-numeric values.")
 
         # try to slice df
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

File: pure/metrics.py [hunk 10]
```diff
@@ -1622,33 +1672,70 @@ def _call_pandas(self, df: pd.DataFrame) -> Dict[str, Any]:
...
+        from pyspark.sql.functions import col, to_date, lit, least, min as min_
+        from sklearn.utils import shuffle
+
+        if 'index' not in df.columns:
+            raise ValueError('PySpark dataframe must contain "index" column to get slices.')
+
+        is_similar = True
+        importance_dict = {}
+
+        # check for existence of numeric values
```
[the remark this answers is not in the document]
Comment (dvoykov) [reply pure/metrics.py]: done.

**Pull request merged**